### PR TITLE
feat: SMTP as a fallback function when Sendgrid quota exceeds limit

### LINKED ISSUE #5958: FallBack to SMTP when SendGrid Fails/Exceeds Quota
**Describe the feature request**
We need a feature in the system where the SMTP protocol must be used when SendGrid's quota exceeds/ fails to work. 
This is essential as whenever someone orders a ticket, they expect the tickets/invoice to reach them via mail. A backup service like this would be useful.

**Additional context**
Working on it.

### PR DESCRIPTION
**Pull request by mrsaicharan1**
Fixes #5958 

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The unit tests pass locally with my changes 
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

- [x] All the functions created/modified in this PR contain relevant docstrings.

#### Short description of what this resolves:
- creates a centralized task rather than having two separate tasks for them
- Checks for Sendgrid quota value and decides the protocol

#### Changes proposed in this pull request:

- Added centralised task `send_email_via_smtp`
-  Add check for SMTP fallback

### CODE REVIEW

**Comment by mrsaicharan1** [comment]
The hound violations seem to be out of context.

**Comment by mrsaicharan1** [comment]
@shreyanshdwivedi @uds5501 @iamareebjamal 
Here is the response for sendgrid quota which I've implemented
`HTTP/1.1 200
[
  {
    "created_at": 1451498784,
    "email_to": "saicharan.reddy1@gmail.com",
    "id": 46,
    "percentage": 90,
    "type": "usage_limit",
    "updated_at": 1451498784
  }]
`

File: app/api/helpers/mail.py [hunk 1]
```diff
@@ -19,6 +20,7 @@ def send_email(to, action, subject, html, attachments=None):
     """
     Sends email and records it in DB
     """
+    from .tasks import send_email_task
```
Comment (uds5501) [inline app/api/helpers/mail.py]: Try importing this at the beginning? maybe that'll help with hound

**Comment by iamareebjamal** [comment]
@mrsaicharan1 Instead of checking sendgrid and sending mail, please send the mail and handle the error if quota is exceeded

**Comment by iamareebjamal** [comment]
Even before so, the implementation should be that it reads config and only tries to use sendgrid if configured, else fallback to SMTP

**Comment by mrsaicharan1** [comment]
@iamareebjamal I can't understand why Travis is throwing `python_http_client.exceptions.UnauthorizedError: HTTP Error 401: Unauthorized
`
Can you please help me out here?

**Comment by iamareebjamal** [comment]
Because it is trying to send the email and `Sendgrid key not defined` in Travis build environment

If you ask me, it is perfect. You implement the SMTP fallback correclty and build will automatically pass. TDD

**Comment by mrsaicharan1** [comment]
@iamareebjamal Please check now.

**Comment by mrsaicharan1** [comment]
@iamareebjamal Please check

**Comment by mrsaicharan1** [comment]
@iamareebjamal Travis seems to fail. `ERROR:marrow.mailer:Delivery of message <155978267199.7284.14634174120944382671@travis-job-28221724-4b92-49d7-9b81-ed85240b5743> failed.` Is this an internal config error? Locally, I've implemented a sample mailing function which is similar to the one in our system. I was able to send mails successfully.

**Comment by iamareebjamal** [comment]
Because SMTP is not configured in travis. You need to not send the email if it is not configured and just log the error. It was happening in past implementation

**Comment by mrsaicharan1** [comment]
@iamareebjamal Got it.

**Comment by mrsaicharan1** [comment]
@iamareebjamal I've logged the errors. But they it still seems to fail.

**Comment by iamareebjamal** [comment]
You should not send the mail if SMTP setiings are not configured

**Comment by mrsaicharan1** [comment]
@iamareebjamal I've handled the errors properly but I don't know what's going wrong 
```
ERROR:root:SendGrid Key not found....Falling back to SMTP
Traceback (most recent call last):
  File "/home/travis/build/fossasia/open-event-server/app/api/helpers/mail.py", line 85, in send_email
    "Authorization": ("Bearer " + key),
TypeError: must be str, not NoneType
ERROR:marrow.mailer:Delivery of message <155993309409.7296.2050613144652800737@travis-job-743d757a-1fc7-49fe-b33f-2f39751278dc> failed.
ERROR:root:Error in sending mail via SMTP
Traceback (most recent call last):
  File "/home/travis/build/fossasia/open-event-server/app/api/helpers/mail.py", line 85, in send_email
    "Authorization": ("Bearer " + key),
TypeError: must be str, not NoneType

```

**Comment by mrsaicharan1** [comment]
@uds5501 @shreyanshdwivedi Just a background, whenever an HTTP error code 429 (Too many Requests) is thrown, we shift to SMTP as an alternate path. 
Screenshot 2019-06-07 at 3 33 48 PM

**Comment by mrsaicharan1** [comment]
@uds5501 @shreyanshdwivedi Please review this.

**Comment by mrsaicharan1** [comment]
@shreyanshdwivedi @uds5501 @iamareebjamal Please review.

File: app/api/helpers/mail.py [hunk 2]
```diff
@@ -35,40 +54,53 @@ def send_email(to, action, subject, html, attachments=None):
...
+
+                        except requests.exceptions.HTTPError as e:
+                            if e.code == 429:
+                                logging.exception("Sendgrid quota has exceeded")
+                                send_email_task.delay(payload=payload, headers=None, protocol='smtp', config=config)
+
+                        except Exception as e:
+                            logging.exception("Error while sending email-{}".format(str(e)))
+
+                    except TypeError:
```
Comment (iamareebjamal) [inline app/api/helpers/mail.py]: Don't use exceptions for non-exceptional behaviour.

```python
key = get_settings().get('sendgrid_key')
if key:
    # great
else:
    # sad
```

File: app/api/helpers/mail.py [hunk 3]
```diff
@@ -35,40 +54,53 @@ def send_email(to, action, subject, html, attachments=None):
...
+                        key = get_settings()['sendgrid_key']
+                        headers = {
+                            "Authorization": ("Bearer " + key),
+                            "Content-Type": "application/json"
+                        }
+                        try:
+                            payload['fromname'] = email_from_name
+                            send_email_task.delay(payload=payload, headers=headers, protocol='sendgrid')
+
+                        except requests.exceptions.HTTPError as e:
```
Comment (iamareebjamal) [inline app/api/helpers/mail.py]: I'm confident this will never get called since you are using `.delay` above

File: app/api/helpers/mail.py [hunk 4]
```diff
@@ -35,40 +54,53 @@ def send_email(to, action, subject, html, attachments=None):
...
+                'password': get_settings()['smtp_password'],
+                'encryption': smtp_encryption,
+                'port': get_settings()['smtp_port'],
+            }
+            smtp_status = check_smtp_config(smtp_encryption)
+            if smtp_status:
+                if email_service == 'smtp':
+                    try:
+                        send_email_task.delay(payload=payload, headers=None, protocol='smtp', config=config)
+                    except Exception as e:
```
Comment (iamareebjamal) [inline app/api/helpers/mail.py]: Same here

File: app/api/helpers/tasks.py [hunk 5]
```diff
@@ -52,55 +52,56 @@
...
-            attachment.file_content = FileContent(encoded)
-            attachment.file_type = FileType('application/pdf')
-            attachment.file_name = FileName(payload['to'])
-            attachment.disposition = Disposition('attachment')
-            message.add_attachment(attachment)
-    sendgrid_client = SendGridAPIClient(get_settings()['sendgrid_key'])
-    logging.info('Sending an email regarding {} on behalf of {}'.format(payload["subject"], payload["from"]))
-    try:
+def send_email_task(payload, headers, protocol, config=None):
+    if protocol == 'sendgrid':
```
Comment (iamareebjamal) [inline app/api/helpers/tasks.py]: Why complicate things?

Create 2 functions, one for smtp, one for sendgrid

File: app/api/helpers/mail.py (hunk 3 above)
Comment (mrsaicharan1) [reply app/api/helpers/mail.py]: @iamareebjamal I think its better to handle the exceptions in the task functions itself. Will update.

File: app/api/helpers/tasks.py (hunk 5 above)
Comment (mrsaicharan1) [reply app/api/helpers/tasks.py]: Yeah, my original idea was to create a centralised function but this seems easier. Will update.

File: app/api/helpers/mail.py [hunk 6]
```diff
@@ -35,40 +54,44 @@ def send_email(to, action, subject, html, attachments=None):
...
+            }
+            smtp_status = check_smtp_config(smtp_encryption)
+            if smtp_status:
+                if email_service == 'smtp':
+                    try:
+                        send_email_task_smtp.delay(payload=payload, headers=None, config=config)
+                    except Exception as e:
+                        logging.exception("Error while sending email-{}".format(str(e)))
+                else:
+                    key = get_settings()['sendgrid_key']
```
Comment (iamareebjamal) [inline app/api/helpers/mail.py]: It'll throw KeyError, I wrote `get_settings().get('sendgrid_key')`

File: app/api/helpers/mail.py (hunk 4 above)
Comment (iamareebjamal) [reply app/api/helpers/mail.py]: Still not resolved

File: app/api/helpers/tasks.py [hunk 7]
```diff
@@ -51,56 +51,66 @@
...
+        message.plain = strip_tags(payload['html'])
+        message.rich = payload['html']
+        if payload['attachments'] is not None:
+            for attachment in payload['attachments']:
+                message.attach(name=attachment)
+        mailer.send(message)
+        logging.info('Message sent via SMTP')
+        mailer.stop()
+    except Exception:
+        logging.exception('Error in sending mail via SMTP')
```
Comment (iamareebjamal) [inline app/api/helpers/tasks.py]: Unless you do something in the block, generic exception catching is useless

File: app/api/helpers/tasks.py [hunk 8]
```diff
@@ -51,56 +51,66 @@
...
         sendgrid_client.send(message)
         logging.info('Email sent successfully')
+    except requests.exceptions.HTTPError as e:
+        if e.code == 429:
+            logging.exception("Sendgrid quota has exceeded")
+        else:
+            logging.exception("The following error has occurred with sendgrid-{}".format(str(e)))
     except Exception:
-        logging.exception('Error occured while sending the email')
+        logging.exception('Error in sending mail via Sendgrid')
```
Comment (iamareebjamal) [inline app/api/helpers/tasks.py]: Same here

File: app/api/helpers/tasks.py (hunk 7 above)
Comment (mrsaicharan1) [reply app/api/helpers/tasks.py]: @iamareebjamal Will catching a specific exception and logging the error message do? Like `DeliveryFailedException` & `MessageFailedException`?

Comment (iamareebjamal) [reply app/api/helpers/tasks.py]: Why do you need to catch that. Let it be thrown

File: app/api/helpers/tasks.py [hunk 9]
```diff
@@ -7,6 +7,7 @@
 from flask import current_app, render_template
 from flask_celeryext import RequestContextTask
 from marrow.mailer import Mailer, Message
+from marrow.mailer.exc import MessageFailedException, DeliveryFailedException
```
Comment (Anupam-dagar) [inline app/api/helpers/tasks.py]: You have unused imports.

Comment (mrsaicharan1) [reply app/api/helpers/tasks.py]: Yeah, I'm focusing on fixing travis first. Will do it.

**Comment by mrsaicharan1** [comment]
@iamareebjamal I've made the changes. But the other tests which are related to sending an email are failing

```
thod: POST
uri: /v1/auth/change-password
headers: 
    Content-Type: application/json
    Authorization: JWT eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJleHAiOjE1NjA2OTExNDcsImlhdCI6MTU2MDYwNDc0NywibmJmIjoxNTYwNjA0NzQ3LCJpZGVudGl0eSI6MX0.hgCpfAlLq3PcgMFza2QtUgxeDgfbeaI_dpj3tCmD8ZA
    User-Agent: Dredd/4.0.0 (Linux 4.4.0-101-generic; x64)
    Content-Length: 95
body: 
{
  "data": {
    "old-password": "fossasia",
    "new-password": "eventyay"
  }
}
expected: 
headers: 
    Content-Type: application/json
body: 
{
  "email": "example@example.com",
  "id": "1",
  "name": "example",
  "password-changed": true
}
statusCode: 200
actual: 
statusCode: 500
headers: 
    content-type: application/vnd.api+json
    content-length: 140
    access-control-allow-origin: *
    server: Werkzeug/0.15.4 Python/3.6.3
    date: Sat, 15 Jun 2019 13:31:13 GMT
body: 
{"errors": [{"detail": "Unknown error", "source": {"pointer": ""}, "status": 500, "title": "Unknown error"}], "jsonapi": {"version": "1.0"}}

```

**Comment by iamareebjamal** [comment]
Because you are still sending emails when smtp is not configured. The send_email_via_smtp method should not be simply called if SMTP settings are not configured

**Comment by mrsaicharan1** [comment]
Can you point where I'm calling it when it's not configured? I can't find
it.

**Comment by iamareebjamal** [comment]
Then there's some internal error in a case when neither sendgrid, nor smtp is configured. Replicate the environment and try the same URL

**Comment by mrsaicharan1** [comment]
@iamareebjamal Tried all the tests failing on travis locally. They were working fine.
@shreyanshdwivedi @uds5501 @CosmicCoder96 Could you take a look as well once?

**Comment by iamareebjamal** [comment]
In your local environment, is SMTP or sendgrid configured?

**Comment by mrsaicharan1** [comment]
> In your local environment, is SMTP or sendgrid configured?

I removed both to test it.

**Comment by iamareebjamal** [comment]
If we had this fixed https://github.com/fossasia/open-event-server/issues/4903, we wouldn't have to guess what's wrong

**Comment by mrsaicharan1** [comment]
> If we had this fixed #4903, we wouldn't have to guess what's wrong

I agree. It just throws 500 errors without specifying what the error is :/.

**Comment by iamareebjamal** [comment]
I'll try reproducing it today if I get time hopefully

**Comment by mrsaicharan1** [comment]
Yeah, that would be great.

**Comment by iamareebjamal** [comment]
@mrsaicharan1 

Tests are failing because your code is trying to send emails and failing:

```
 File "/home/travis/build/fossasia/open-event-server/app/api/auth.py", line 225, in reset_password_post
    send_email_with_action(user, PASSWORD_RESET, app_name=get_settings()['app_name'], link=link)
  File "/home/travis/build/fossasia/open-event-server/app/api/helpers/mail.py", line 119, in send_email_with_action
    html=MAILS[action]['message'].format(**kwargs)
  File "/home/travis/build/fossasia/open-event-server/app/api/helpers/mail.py", line 87, in send_email
    send_email_task_sendgrid.delay(payload=payload, headers=headers)
...
File "/home/travis/build/fossasia/open-event-server/app/api/helpers/tasks.py", line 76, in send_email_task_sendgrid
    sendgrid_client.send(message)
  File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/sendgrid/sendgrid.py", line 98, in send
    response = self.client.mail.send.post(request_body=message.get())
  File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/python_http_client/client.py", line 252, in http_request
    return Response(self._make_request(opener, request, timeout=timeout))
  File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/python_http_client/client.py", line 176, in _make_request
    raise exc
python_http_client.exceptions.UnauthorizedError: HTTP Error 401: Unauthorized
```

**Comment by iamareebjamal** [comment]
Found the freaking culprit: https://api.eventyay.com/#settings-settings-details-patch

Not only it sets the sendgrid key to a random value, it also changes the environment to production!!!

Fix the docs and the error will be gone

**Comment by iamareebjamal** [comment]
Something else is changing the app config. Need to debug more

**Comment by mrsaicharan1** [comment]
@iamareebjamal This is only due to the random app config change right? Because I made sure that the email is not being sent when neither of them are configured.

**Comment by iamareebjamal** [comment]
Yes, it is due to config changes

**Comment by mrsaicharan1** [comment]
@iamareebjamal Looks good I guess. Travis seems to go through now :)

File: app/api/helpers/tasks.py (hunk 9 above)
Comment (iamareebjamal) [reply app/api/helpers/tasks.py]: Focus now

File: app/api/helpers/tasks.py [hunk 10]
```diff
@@ -51,58 +52,62 @@
...
+                message.add_attachment(attachment)
+        sendgrid_client = SendGridAPIClient(get_settings()['sendgrid_key'])
+        logging.info('Sending an email regarding {} on behalf of {}'.format(payload["subject"], payload["from"]))
         sendgrid_client.send(message)
         logging.info('Email sent successfully')
-    except Exception:
-        logging.exception('Error occured while sending the email')
+    except urllib.error.HTTPError as e:
+        if e.code == 429:
+            logging.exception("Sendgrid quota has exceeded")
```
Comment (iamareebjamal) [inline app/api/helpers/tasks.py]: Quota has exceeded, so fall back to SMTP

Comment (iamareebjamal) [inline app/api/helpers/tasks.py]: Change to logger.warning

**Comment by mrsaicharan1** [comment]
@iamareebjamal Please have a look now.

**Pull request merged**